test/controllers/users_controller_test.rb

```diff
@@ -344,27 +344,13 @@ def test_save_referer_params
   end
 
   def test_logout_without_referer
-    get :logout
-    assert_response :success
-    assert_template :logout
-    assert_select "input[name=referer][value=?]", ""
-
-    session_id = assert_select("input[name=session]").first["value"]
-
-    get :logout, :params => { :session => session_id }
```

Comment:
I'm inclined to leave in the get test in method which checks that GET returns the login page - obviously get rid of the assertion that the session ID is in the returned form.

I know you're doing it in the system tests as well, but the cases here do check extra things like that the fallback page preserves the referer?

Reply:
Sure, I've added those bits of the tests back in. I've also added a system test that clicks the button and checks the redirect, as well as a system test that checks the referer works for normal behaviour.

(I added the last two as system tests, since I dislike writing controller tests where you check the form, and then try submitting a facsimile of what you hope the browser sends)